cmd/traffic/cmd/manager/internal/state/intercept.go

```diff
@@ -355,10 +341,68 @@ func waitForConfigMapUpdate(ctx context.Context, cmAPI typed.ConfigMapInterface,
 	}
 }
 
-func (s *State) waitForAgent(ctx context.Context, name, namespace string) error {
+func watchFailedInjectionEvents(ctx context.Context, name, namespace string) (<-chan *events.Event, error) {
+	// A timestamp with second granularity is needed here, because that's what the event creation time uses.
+	// Finer granularity will result in relevant events seemingly being created before this timestamp because
+	// they have the fraction of seconds trimmed off (which is odd, given that the type used is a MicroTime).
+	start := time.Unix(time.Now().Unix(), 0)
+
+	ei := k8sapi.GetK8sInterface(ctx).EventsV1().Events(namespace)
+	w, err := ei.Watch(ctx, meta.ListOptions{
+		FieldSelector: fields.OneTermNotEqualSelector("type", "Normal").String()})
+	if err != nil {
+		return nil, err
+	}
+	nd := name + "-"
+	ec := make(chan *events.Event)
+	go func() {
+		defer w.Stop()
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case eo, ok := <-w.ResultChan():
+				if !ok {
+					return
+				}
+				// Using a negated Before when comparing the timestamps here is relevant. They will often be equal and still relevant
+				if e, ok := eo.Object.(*events.Event); ok &&
+					!e.CreationTimestamp.Time.Before(start) &&
+					!strings.HasPrefix(e.Note, "(combined from similar events):") {
+					n := e.Regarding.Name
+					if strings.HasPrefix(n, nd) || n == name {
+						dlog.Errorf(ctx, "%s", e.Note)
+						ec <- e
+					}
+				}
+			}
+		}
+	}()
+	return ec, nil
+}
+
+func (s *State) waitForAgent(ctx context.Context, name, namespace string, failedCreateCh <-chan *events.Event) error {
 	snapshotCh := s.WatchAgents(ctx, nil)
 	for {
 		select {
+		case fe, ok := <-failedCreateCh:
+			if ok {
+				msg := fe.Note
+				switch fe.Reason {
+				case "Unhealthy":
+					// Let readiness probe continue, this isn't fatal
+					continue
+				case "BackOff":
+					// The traffic-agent container was injected, but it fails to start
+					msg = fmt.Sprintf("%s\nThe logs of %s %s might provide more details", msg, fe.Regarding.Kind, fe.Regarding.Name)
+				case "FailedCreate", "FailedScheduling":
+					// The injection of the traffic-agent failed for some reason, most likely due to resource quota restrictions.
```

Comment:
I feel like the "most likely" part of this comment should be reflected in the message sent to the user unless we are actually certain that a resource quota restriction was the cause of the issue.

Reply:
It should. I made the hint a bit more descriptive.